events/listener.go

```diff
@@ -0,0 +1,675 @@
+	if err != nil {
+		return nil, fmt.Errorf("failed to create BookingToken instance: %w", err)
+	}
+	el.btContracts[addr] = bt
+	return bt, nil
+}
+
+// unsubscribe removes a subscription from the subscriptions map and closes the event channel
+func (el *EventListener) unsubscribe(subID string) {
+	el.mu.Lock()
+	defer el.mu.Unlock()
+	if subInfo, exists := el.subscriptions[subID]; exists {
+		// Cancel the context to stop resubscription
+		subInfo.unsubscribe()
+		// Unsubscribe from the event
+		if subInfo.sub != nil {
+			subInfo.sub.Unsubscribe()
+		}
+		// Close the event channel
+		if subInfo.closeChan != nil {
+			subInfo.closeChan()
+		}
+		delete(el.subscriptions, subID)
+		el.logger.Debugf("Unsubscribed subID %s", subID)
+	}
+}
+
+// listenerHandle is a handle to stop event listeners
+type listenerHandle struct {
+	unsubscribe func()
+}
+
+// Stop stops the event listener
+func (h *listenerHandle) Unsubscribe() {
+	h.unsubscribe()
+}
+
+// RegisterServiceAddedHandler registers a handler for the ServiceAdded event on a CMAccount
+func (el *EventListener) RegisterServiceAddedHandler(cmAccountAddr common.Address, serviceName []string, handler EventHandler) (ListenerHandle, error) {
+	el.mu.Lock()
+	defer el.mu.Unlock()
+
+	subID := uuid.New().String()
+
+	// Get or create CMAccount instance
+	cmAccount, err := el.getOrCreateCMAccount(cmAccountAddr)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create a context with cancellation
+	ctx, cancel := context.WithCancel(context.Background())
+
+	// Store subscription info
+	subInfo := &subscriptionInfo{
+		subID:       subID,
+		contract:    cmAccountAddr,
+		handler:     handler,
+		unsubscribe: cancel,
+	}
+	el.subscriptions[subID] = subInfo
+
+	// Start the resubscription
+	go el.resubscribeServiceAdded(ctx, subID, cmAccount, serviceName)
+
+	// Return handle to stop listening
+	return &listenerHandle{
+		unsubscribe: func() {
+			el.unsubscribe(subID)
+		},
+	}, nil
+}
+
+// resubscribeServiceAdded handles resubscription for ServiceAdded events
+func (el *EventListener) resubscribeServiceAdded(ctx context.Context, subID string, cmAccount *cmaccount.Cmaccount, serviceName []string) {
+	backoffMax := 2 * time.Minute // Maximum backoff time between retries
+
+	resubscribeFn := func(ctx context.Context, lastError error) (event.Subscription, error) {
+		if lastError != nil {
+			el.logger.Errorf("Resubscribe attempt after error: %v", lastError)
+		}
+
+		eventChan := make(chan *cmaccount.CmaccountServiceAdded)
+
+		// Subscribe to the event
+		sub, err := cmAccount.WatchServiceAdded(&bind.WatchOpts{Context: ctx}, eventChan, serviceName)
+		if err != nil {
+			el.logger.Errorf("Failed to subscribe to ServiceAdded events: %v", err)
+			return nil, err
+		}
+
+		el.mu.Lock()
+		subInfo, exists := el.subscriptions[subID]
+		if !exists {
+			el.mu.Unlock()
+			sub.Unsubscribe()
+			return nil, fmt.Errorf("subscription %s no longer exists", subID)
+		}
+		subInfo.sub = sub
```

Comment:
is it possible that subInfo would be accessed concurrently? not map, but subInfo itself

Reply:
Yup, sounds logical to protect it with a sync. Updated.